<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.10.0" />
<title>bearclaw.io API documentation</title>
<meta name="description" content="" />
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/sanitize.min.css" integrity="sha256-PK9q560IAAa6WVRRh76LtCaI8pjTJ2z11v0miyNNjrs=" crossorigin>
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/typography.min.css" integrity="sha256-7l/o7C8jubJiy74VsKTidCy1yBkRtiUGbVkYBylBqUg=" crossorigin>
<link rel="stylesheet preload" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/styles/github.min.css" crossorigin>
<style>:root{--highlight-color:#fe9}.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}#sidebar > *:last-child{margin-bottom:2cm}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}h1:target,h2:target,h3:target,h4:target,h5:target,h6:target{background:var(--highlight-color);padding:.2em 0}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{margin-top:.6em;font-weight:bold}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}dt:target .name{background:var(--highlight-color)}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}td{padding:0 .5em}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%;height:100vh;overflow:auto;position:sticky;top:0}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
<script async src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS_CHTML" integrity="sha256-kZafAc6mZvK3W3v1pHOcUix30OHQN6pU/NO2oFkqZVw=" crossorigin></script>
<script defer src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/highlight.min.js" integrity="sha256-Uv3H6lx7dJmRfRvH8TH6kJD1TSK1aFcwgx+mdg3epi8=" crossorigin></script>
<script>window.addEventListener('DOMContentLoaded', () => hljs.initHighlighting())</script>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>bearclaw.io</code></h1>
</header>
<section id="section-intro">
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">from functools import wraps
from pathlib import Path
from shutil import copyfileobj, copyfile
from tempfile import TemporaryDirectory
import gzip

from SigProfilerMatrixGenerator.scripts import SigProfilerMatrixGeneratorFunc as matGen
from pandas import DataFrame, Series

from bearclaw.io import tso, vcf
from bearclaw.feature_extraction import _get_cosmic_feature_names


def from_tso500_trace(method):
    &#34;&#34;&#34;Decorate method with a TSO500 TMB trace to VCF conversion preprocessing step.&#34;&#34;&#34;

    @wraps(method)
    def decorated_transform(input_trace_tsv: Path, *args, **kwargs):
        with TemporaryDirectory() as tmp_dir:
            output_vcf = Path(tmp_dir) / input_trace_tsv.name.replace(&#34;tsv&#34;, &#34;vcf&#34;)
            tso.tmb_trace_to_vcf(input_trace_tsv, output_vcf)
            x_transformed = method(output_vcf, *args, **kwargs)
        return x_transformed

    return decorated_transform


def extract_mutation_spectra(input_vcf: Path, exome: bool = True) -&gt; dict[str, Series]:
    &#34;&#34;&#34;Classify variants according to mutation type and pool per sample and type.

    Uses SigProfilerMatrixGenerator [1] with GRCh37 to extract the following
    classes:
    - 96 single base substitutions,
    - 78 doublet base substitutions,
    - 83 indels.

    Args:
        input_vcf: A VCF file with variants used to construct spectrum.

    Returns: Dictionary of series with number of mutations per mutation type,
        where the keys are the classes: `single_base_substitutions`,
        `doublet_base_substitutions`, and `indel`.

    [1]: Bergstrom EN, Huang MN, Mahto U, Barnes M, Stratton MR, Rozen SG, and
        Alexandrov LB (2019) SigProfilerMatrixGenerator: a tool for visualizing
        and exploring patterns of small mutational events. BMC Genomics 20,
        Article number: 685.
    &#34;&#34;&#34;
    # Provision temporary folder to run SigProfilerMatrixGenerator.
    with TemporaryDirectory() as tmp_dir:
        # Decompress VCF file for SigProfilerMatrixGenerator.
        if input_vcf.name.endswith(&#34;gz&#34;):
            target_vcf = Path(tmp_dir) / input_vcf.name.removesuffix(&#34;.gz&#34;)
            with gzip.open(input_vcf, &#34;rb&#34;) as fo_gz:
                with open(target_vcf, &#34;wb&#34;) as fo_out:
                    copyfileobj(fo_gz, fo_out)
        else:
            target_vcf = Path(tmp_dir) / input_vcf.name
            copyfile(str(input_vcf), str(target_vcf))

        # Unique project name.
        name = &#34;mutation-class&#34;
        matrices = matGen.SigProfilerMatrixGeneratorFunc(
            project=name,
            genome=&#34;GRCh37&#34;,
            vcfFiles=tmp_dir,
            plot=False,
            exome=exome,
            seqInfo=False,
        )

    index = input_vcf.name.split(&#34;.&#34;)[0]

    sbs96_columns = _get_cosmic_feature_names(
        &#34;single_base_substitutions&#34;, signatures=False
    )
    db78_columns = _get_cosmic_feature_names(
        &#34;doublet_base_substitutions&#34;, signatures=False
    )
    indel_columns = _get_cosmic_feature_names(&#34;indel&#34;, signatures=False)
    X_SBS96 = DataFrame(0, columns=sbs96_columns, index=[index])
    X_DB78 = DataFrame(0, columns=db78_columns, index=[index])
    X_ID83 = DataFrame(0, columns=indel_columns, index=[index])
    if &#34;96&#34; in matrices:
        X_SBS96 = matrices[&#34;96&#34;].transpose()
    if &#34;DINUC&#34; in matrices:
        X_DB78 = matrices[&#34;DINUC&#34;].transpose()
    if &#34;ID&#34; in matrices:
        X_ID83 = matrices[&#34;ID&#34;].transpose()

    return {
        &#34;single_base_substitutions&#34;: X_SBS96,
        &#34;doublet_base_substitutions&#34;: X_DB78,
        &#34;indel&#34;: X_ID83,
    }</code></pre>
</details>
</section>
<section>
<h2 class="section-title" id="header-submodules">Sub-modules</h2>
<dl>
<dt><code class="name"><a title="bearclaw.io.tso" href="tso.html">bearclaw.io.tso</a></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt><code class="name"><a title="bearclaw.io.vcf" href="vcf.html">bearclaw.io.vcf</a></code></dt>
<dd>
<div class="desc"></div>
</dd>
</dl>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-functions">Functions</h2>
<dl>
<dt id="bearclaw.io.extract_mutation_spectra"><code class="name flex">
<span>def <span class="ident">extract_mutation_spectra</span></span>(<span>input_vcf: pathlib.Path, exome: bool = True) ‑> dict[str, pandas.core.series.Series]</span>
</code></dt>
<dd>
<div class="desc"><p>Classify variants according to mutation type and pool per sample and type.</p>
<p>Uses SigProfilerMatrixGenerator [1] with GRCh37 to extract the following
classes:
- 96 single base substitutions,
- 78 doublet base substitutions,
- 83 indels.</p>
<h2 id="args">Args</h2>
<dl>
<dt><strong><code>input_vcf</code></strong></dt>
<dd>A VCF file with variants used to construct spectrum.</dd>
</dl>
<p>Returns: Dictionary of series with number of mutations per mutation type,
where the keys are the classes: <code>single_base_substitutions</code>,
<code>doublet_base_substitutions</code>, and <code>indel</code>.</p>
<p>[1]: Bergstrom EN, Huang MN, Mahto U, Barnes M, Stratton MR, Rozen SG, and
Alexandrov LB (2019) SigProfilerMatrixGenerator: a tool for visualizing
and exploring patterns of small mutational events. BMC Genomics 20,
Article number: 685.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def extract_mutation_spectra(input_vcf: Path, exome: bool = True) -&gt; dict[str, Series]:
    &#34;&#34;&#34;Classify variants according to mutation type and pool per sample and type.

    Uses SigProfilerMatrixGenerator [1] with GRCh37 to extract the following
    classes:
    - 96 single base substitutions,
    - 78 doublet base substitutions,
    - 83 indels.

    Args:
        input_vcf: A VCF file with variants used to construct spectrum.

    Returns: Dictionary of series with number of mutations per mutation type,
        where the keys are the classes: `single_base_substitutions`,
        `doublet_base_substitutions`, and `indel`.

    [1]: Bergstrom EN, Huang MN, Mahto U, Barnes M, Stratton MR, Rozen SG, and
        Alexandrov LB (2019) SigProfilerMatrixGenerator: a tool for visualizing
        and exploring patterns of small mutational events. BMC Genomics 20,
        Article number: 685.
    &#34;&#34;&#34;
    # Provision temporary folder to run SigProfilerMatrixGenerator.
    with TemporaryDirectory() as tmp_dir:
        # Decompress VCF file for SigProfilerMatrixGenerator.
        if input_vcf.name.endswith(&#34;gz&#34;):
            target_vcf = Path(tmp_dir) / input_vcf.name.removesuffix(&#34;.gz&#34;)
            with gzip.open(input_vcf, &#34;rb&#34;) as fo_gz:
                with open(target_vcf, &#34;wb&#34;) as fo_out:
                    copyfileobj(fo_gz, fo_out)
        else:
            target_vcf = Path(tmp_dir) / input_vcf.name
            copyfile(str(input_vcf), str(target_vcf))

        # Unique project name.
        name = &#34;mutation-class&#34;
        matrices = matGen.SigProfilerMatrixGeneratorFunc(
            project=name,
            genome=&#34;GRCh37&#34;,
            vcfFiles=tmp_dir,
            plot=False,
            exome=exome,
            seqInfo=False,
        )

    index = input_vcf.name.split(&#34;.&#34;)[0]

    sbs96_columns = _get_cosmic_feature_names(
        &#34;single_base_substitutions&#34;, signatures=False
    )
    db78_columns = _get_cosmic_feature_names(
        &#34;doublet_base_substitutions&#34;, signatures=False
    )
    indel_columns = _get_cosmic_feature_names(&#34;indel&#34;, signatures=False)
    X_SBS96 = DataFrame(0, columns=sbs96_columns, index=[index])
    X_DB78 = DataFrame(0, columns=db78_columns, index=[index])
    X_ID83 = DataFrame(0, columns=indel_columns, index=[index])
    if &#34;96&#34; in matrices:
        X_SBS96 = matrices[&#34;96&#34;].transpose()
    if &#34;DINUC&#34; in matrices:
        X_DB78 = matrices[&#34;DINUC&#34;].transpose()
    if &#34;ID&#34; in matrices:
        X_ID83 = matrices[&#34;ID&#34;].transpose()

    return {
        &#34;single_base_substitutions&#34;: X_SBS96,
        &#34;doublet_base_substitutions&#34;: X_DB78,
        &#34;indel&#34;: X_ID83,
    }</code></pre>
</details>
</dd>
<dt id="bearclaw.io.from_tso500_trace"><code class="name flex">
<span>def <span class="ident">from_tso500_trace</span></span>(<span>method)</span>
</code></dt>
<dd>
<div class="desc"><p>Decorate method with a TSO500 TMB trace to VCF conversion preprocessing step.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def from_tso500_trace(method):
    &#34;&#34;&#34;Decorate method with a TSO500 TMB trace to VCF conversion preprocessing step.&#34;&#34;&#34;

    @wraps(method)
    def decorated_transform(input_trace_tsv: Path, *args, **kwargs):
        with TemporaryDirectory() as tmp_dir:
            output_vcf = Path(tmp_dir) / input_trace_tsv.name.replace(&#34;tsv&#34;, &#34;vcf&#34;)
            tso.tmb_trace_to_vcf(input_trace_tsv, output_vcf)
            x_transformed = method(output_vcf, *args, **kwargs)
        return x_transformed

    return decorated_transform</code></pre>
</details>
</dd>
</dl>
</section>
<section>
</section>
</article>
<nav id="sidebar">
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="bearclaw" href="../index.html">bearclaw</a></code></li>
</ul>
</li>
<li><h3><a href="#header-submodules">Sub-modules</a></h3>
<ul>
<li><code><a title="bearclaw.io.tso" href="tso.html">bearclaw.io.tso</a></code></li>
<li><code><a title="bearclaw.io.vcf" href="vcf.html">bearclaw.io.vcf</a></code></li>
</ul>
</li>
<li><h3><a href="#header-functions">Functions</a></h3>
<ul class="">
<li><code><a title="bearclaw.io.extract_mutation_spectra" href="#bearclaw.io.extract_mutation_spectra">extract_mutation_spectra</a></code></li>
<li><code><a title="bearclaw.io.from_tso500_trace" href="#bearclaw.io.from_tso500_trace">from_tso500_trace</a></code></li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc" title="pdoc: Python API documentation generator"><cite>pdoc</cite> 0.10.0</a>.</p>
</footer>
</body>
</html>