<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.10.0" />
<title>bearclaw.io.vcf API documentation</title>
<meta name="description" content="" />
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/sanitize.min.css" integrity="sha256-PK9q560IAAa6WVRRh76LtCaI8pjTJ2z11v0miyNNjrs=" crossorigin>
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/typography.min.css" integrity="sha256-7l/o7C8jubJiy74VsKTidCy1yBkRtiUGbVkYBylBqUg=" crossorigin>
<link rel="stylesheet preload" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/styles/github.min.css" crossorigin>
<style>:root{--highlight-color:#fe9}.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}#sidebar > *:last-child{margin-bottom:2cm}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}h1:target,h2:target,h3:target,h4:target,h5:target,h6:target{background:var(--highlight-color);padding:.2em 0}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{margin-top:.6em;font-weight:bold}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}dt:target .name{background:var(--highlight-color)}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}td{padding:0 .5em}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%;height:100vh;overflow:auto;position:sticky;top:0}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
<script async src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS_CHTML" integrity="sha256-kZafAc6mZvK3W3v1pHOcUix30OHQN6pU/NO2oFkqZVw=" crossorigin></script>
<script defer src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/highlight.min.js" integrity="sha256-Uv3H6lx7dJmRfRvH8TH6kJD1TSK1aFcwgx+mdg3epi8=" crossorigin></script>
<script>window.addEventListener('DOMContentLoaded', () => hljs.initHighlighting())</script>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>bearclaw.io.vcf</code></h1>
</header>
<section id="section-intro">
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">from functools import wraps
from pathlib import Path
import re
from tempfile import TemporaryDirectory
import warnings

import vcfpy
from vcfpy.exceptions import FieldInfoNotFound, VCFPyException


def non_synonymous(method):
    &#34;&#34;&#34;Decorator that pre-filters synonymous variants from the VCF file.&#34;&#34;&#34;

    @wraps(method)
    def decorated_transform(input_vcf: Path, *args, **kwargs):
        with TemporaryDirectory() as tmp_dir:
            target_vcf = Path(tmp_dir) / input_vcf.name.replace(&#34;.gz&#34;, &#34;&#34;)
            filter_synonymous_variants(input_vcf, target_vcf, soft=False)
            x_transformed = method(target_vcf, *args, **kwargs)
        return x_transformed

    return decorated_transform


def _is_non_synonymous(annotation: dict) -&gt; bool:
    &#34;&#34;&#34;Test if an annotated variant is non-synonymous.

    Args:
        annotation: dict where keys and values are according to ANN standard.
    &#34;&#34;&#34;
    field_names = tuple(annotation.keys())

    # The standard for the annotation field can be found here:
    # https://pcingola.github.io/SnpEff/adds/VCFannotationformat_v1.0.pdf
    annotation_field = field_names[1]
    hgsvp_field = field_names[10]
    # Only variants affecting protein.
    if annotation[hgsvp_field] != &#34;&#34;:
        # Only variants that change amino-acid sequence.
        var_annot_values = annotation[annotation_field].split(&#34;&amp;&#34;)
        synonymous_type = (
            &#34;synonymous_variant&#34;,
            &#34;stop_retained_variant&#34;,
            &#34;start_retained_variant&#34;,
        )
        is_non_synonymous = all(
            v.strip() not in synonymous_type for v in var_annot_values
        )
        if is_non_synonymous:
            return True
    return False


def _check_vcf_ann_header(reader: vcfpy.Reader):
    &#34;&#34;&#34;Validate header to see if INFO column has variant annotation `ANN` fields.

    Raises:
        VCFPyException: When header specifies no ANN field  or incorrect ANN keys.
    &#34;&#34;&#34;
    with warnings.catch_warnings():
        warnings.filterwarnings(&#34;error&#34;)
        try:
            annotation = reader.header.get_info_field_info(&#34;ANN&#34;)
        except FieldInfoNotFound:
            raise VCFPyException(&#34;Missing ANN field in VCF.&#34;)

    return annotation


def _get_annotations_keys(reader: vcfpy.Reader) -&gt; list:
    &#34;&#34;&#34;Extract the annotation&#39;s data fields (=keys) from the VCF header.

    This corresponds to the data fields in the value corresponding to the
    `ANN` key in the `INFO`field.

    Returns:
        Keys corresponding to record ANN array.
    &#34;&#34;&#34;
    annotation = _check_vcf_ann_header(reader)
    ann_format = annotation.description
    mo = re.search(&#34;&#39;(.+)&#39;&#34;, ann_format)
    ann_keys = [key.strip() for key in mo.group(1).split(&#34;|&#34;)]

    # According to standard, there are 16 annotations data fields.
    assert len(ann_keys) == 16, &#34;Variant annotations not compatible with standard.&#34;

    return ann_keys


def _get_annotations(record: vcfpy.Record, reader: vcfpy.Reader) -&gt; list[dict]:
    &#34;&#34;&#34;Turn annotations data field into dictionary.

    Args:
        record: Single record in the VCF file.
        reader: VCFpy `Reader` istance with header information.

    Returns:
        Annotation dictionary per allele.
    &#34;&#34;&#34;

    ann_keys = _get_annotations_keys(reader)
    return [
        dict(zip(ann_keys, ann_values.split(&#34;|&#34;))) for ann_values in record.INFO[&#34;ANN&#34;]
    ]


def filter_synonymous_variants(input_vcf: Path, filtered_vcf: Path, soft: bool = False):
    &#34;&#34;&#34;Filter synonymous (non-amino acid sequencing changing) variants.

    Args:
        input_vcf: Variant file (possbily gzipped) to filter.
        filtered_vcf: Store filtered VCF to this (possbily gzipped) file.
        soft: When `True`, apply a soft filter that only changes the FILTER
            column (but keeps the record). When `False`, remove filtered
            records.

    Raises:
        VCFPyException: When header specifies no `ANN` field  or incorrect `ANN`
            keys.
    &#34;&#34;&#34;
    with vcfpy.Reader.from_path(str(input_vcf)) as reader:
        # Soft filter adds the following variant FILTER.
        if soft:
            reader.header.add_filter_line(
                vcfpy.OrderedDict(
                    [
                        (&#34;ID&#34;, &#34;IS_SYNON&#34;),
                        (
                            &#34;Description&#34;,
                            &#34;The variant is not amino-acid sequence changing.&#34;,
                        ),
                    ]
                )
            )

        with vcfpy.Writer.from_path(filtered_vcf, reader.header) as writer:
            for record in reader:
                try:
                    annotations = _get_annotations(record, reader)
                except KeyError:
                    # Filter variants without annotation.
                    if not soft:
                        continue
                    record.add_filter(&#34;IS_SYNON&#34;)
                else:
                    # Filter variants that are not non-synonymous.
                    if not any(map(_is_non_synonymous, annotations)):
                        if not soft:
                            continue
                        record.add_filter(&#34;IS_SYNON&#34;)

                writer.write_record(record)</code></pre>
</details>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-functions">Functions</h2>
<dl>
<dt id="bearclaw.io.vcf.filter_synonymous_variants"><code class="name flex">
<span>def <span class="ident">filter_synonymous_variants</span></span>(<span>input_vcf: pathlib.Path, filtered_vcf: pathlib.Path, soft: bool = False)</span>
</code></dt>
<dd>
<div class="desc"><p>Filter synonymous (non-amino acid sequencing changing) variants.</p>
<h2 id="args">Args</h2>
<dl>
<dt><strong><code>input_vcf</code></strong></dt>
<dd>Variant file (possbily gzipped) to filter.</dd>
<dt><strong><code>filtered_vcf</code></strong></dt>
<dd>Store filtered VCF to this (possbily gzipped) file.</dd>
<dt><strong><code>soft</code></strong></dt>
<dd>When <code>True</code>, apply a soft filter that only changes the FILTER
column (but keeps the record). When <code>False</code>, remove filtered
records.</dd>
</dl>
<h2 id="raises">Raises</h2>
<dl>
<dt><code>VCFPyException</code></dt>
<dd>When header specifies no <code>ANN</code> field
or incorrect <code>ANN</code>
keys.</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def filter_synonymous_variants(input_vcf: Path, filtered_vcf: Path, soft: bool = False):
    &#34;&#34;&#34;Filter synonymous (non-amino acid sequencing changing) variants.

    Args:
        input_vcf: Variant file (possbily gzipped) to filter.
        filtered_vcf: Store filtered VCF to this (possbily gzipped) file.
        soft: When `True`, apply a soft filter that only changes the FILTER
            column (but keeps the record). When `False`, remove filtered
            records.

    Raises:
        VCFPyException: When header specifies no `ANN` field  or incorrect `ANN`
            keys.
    &#34;&#34;&#34;
    with vcfpy.Reader.from_path(str(input_vcf)) as reader:
        # Soft filter adds the following variant FILTER.
        if soft:
            reader.header.add_filter_line(
                vcfpy.OrderedDict(
                    [
                        (&#34;ID&#34;, &#34;IS_SYNON&#34;),
                        (
                            &#34;Description&#34;,
                            &#34;The variant is not amino-acid sequence changing.&#34;,
                        ),
                    ]
                )
            )

        with vcfpy.Writer.from_path(filtered_vcf, reader.header) as writer:
            for record in reader:
                try:
                    annotations = _get_annotations(record, reader)
                except KeyError:
                    # Filter variants without annotation.
                    if not soft:
                        continue
                    record.add_filter(&#34;IS_SYNON&#34;)
                else:
                    # Filter variants that are not non-synonymous.
                    if not any(map(_is_non_synonymous, annotations)):
                        if not soft:
                            continue
                        record.add_filter(&#34;IS_SYNON&#34;)

                writer.write_record(record)</code></pre>
</details>
</dd>
<dt id="bearclaw.io.vcf.non_synonymous"><code class="name flex">
<span>def <span class="ident">non_synonymous</span></span>(<span>method)</span>
</code></dt>
<dd>
<div class="desc"><p>Decorator that pre-filters synonymous variants from the VCF file.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def non_synonymous(method):
    &#34;&#34;&#34;Decorator that pre-filters synonymous variants from the VCF file.&#34;&#34;&#34;

    @wraps(method)
    def decorated_transform(input_vcf: Path, *args, **kwargs):
        with TemporaryDirectory() as tmp_dir:
            target_vcf = Path(tmp_dir) / input_vcf.name.replace(&#34;.gz&#34;, &#34;&#34;)
            filter_synonymous_variants(input_vcf, target_vcf, soft=False)
            x_transformed = method(target_vcf, *args, **kwargs)
        return x_transformed

    return decorated_transform</code></pre>
</details>
</dd>
</dl>
</section>
<section>
</section>
</article>
<nav id="sidebar">
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="bearclaw.io" href="index.html">bearclaw.io</a></code></li>
</ul>
</li>
<li><h3><a href="#header-functions">Functions</a></h3>
<ul class="">
<li><code><a title="bearclaw.io.vcf.filter_synonymous_variants" href="#bearclaw.io.vcf.filter_synonymous_variants">filter_synonymous_variants</a></code></li>
<li><code><a title="bearclaw.io.vcf.non_synonymous" href="#bearclaw.io.vcf.non_synonymous">non_synonymous</a></code></li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc" title="pdoc: Python API documentation generator"><cite>pdoc</cite> 0.10.0</a>.</p>
</footer>
</body>
</html>