<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.10.0" />
<title>bearclaw.preprocessing API documentation</title>
<meta name="description" content="" />
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/sanitize.min.css" integrity="sha256-PK9q560IAAa6WVRRh76LtCaI8pjTJ2z11v0miyNNjrs=" crossorigin>
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/typography.min.css" integrity="sha256-7l/o7C8jubJiy74VsKTidCy1yBkRtiUGbVkYBylBqUg=" crossorigin>
<link rel="stylesheet preload" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/styles/github.min.css" crossorigin>
<style>:root{--highlight-color:#fe9}.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}#sidebar > *:last-child{margin-bottom:2cm}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}h1:target,h2:target,h3:target,h4:target,h5:target,h6:target{background:var(--highlight-color);padding:.2em 0}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{margin-top:.6em;font-weight:bold}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}dt:target .name{background:var(--highlight-color)}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}td{padding:0 .5em}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%;height:100vh;overflow:auto;position:sticky;top:0}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
<script async src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS_CHTML" integrity="sha256-kZafAc6mZvK3W3v1pHOcUix30OHQN6pU/NO2oFkqZVw=" crossorigin></script>
<script defer src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/highlight.min.js" integrity="sha256-Uv3H6lx7dJmRfRvH8TH6kJD1TSK1aFcwgx+mdg3epi8=" crossorigin></script>
<script>window.addEventListener('DOMContentLoaded', () => hljs.initHighlighting())</script>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>bearclaw.preprocessing</code></h1>
</header>
<section id="section-intro">
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">from pathlib import Path
from typing import Callable, Literal, Optional, Union

from pandas import DataFrame, Series, concat


class VariantDataGenerator:
    &#34;&#34;&#34;Transform (unstructured) mutation files to tabular format.&#34;&#34;&#34;

    def __init__(
        self,
        transform: Callable,
    ):
        &#34;&#34;&#34;
        Generate tabular data from files containing variant or copy number calls.

        Args:
            transform: Method that generates a row from a VCF or CNV file.
        &#34;&#34;&#34;
        self.transform = transform

    def flow_from_dataframe(
        self,
        dataframe: DataFrame,
        x_col=&#34;filename&#34;,
        y_col=&#34;class&#34;,
        class_mode: Optional[Literal[&#34;raw&#34;]] = &#34;raw&#34;,
        keep_columns: bool = True,
        coverage_size: Optional[Union[float, str]] = None,
        decimals: Optional[int] = 2,
    ):
        &#34;&#34;&#34;Load dataset by reading VCF files and target label from dataframe.

        Args:
            dataframe: Pandas dataframe with columns pointing to VCF or CNV files.
            class_mode: When None, don&#39;t extract target label (inference mode).
            x_col: Column pointing to location of VCF or CNV file.
            y_col: Target label column.
            class_mode: Return features and labels during training mode (&#34;raw&#34;),
                or return only features during serving (None).
            keep_columns: Use only the features extracted through `x_col` (False) or
                also concatenate other columns in the dataframe after extraction (True).
            coverage_size: Normalise estimates by value (float), column (str), or not at
                all (None). Usually, this value is the size of the genomic region
                [typically in megabases (mb)] covered at sufficient depth to call
                variants.
            decimals: If not None, round to this many decimals.

        Returns: When class_mode is `None` return features `X`, otherwise return a
            pair (X, y) with labels `y`.

        &#34;&#34;&#34;
        if keep_columns:
            to_keep = dataframe.columns.difference([x_col, y_col])

        if coverage_size is None:
            normalisation = Series(1, index=dataframe.index)
        elif isinstance(coverage_size, (float, int)):
            normalisation = Series(coverage_size, index=dataframe.index)
        elif coverage_size in dataframe.columns:
            normalisation = dataframe[coverage_size]
        else:
            raise KeyError(f&#34;Unknown column {coverage_size}.&#34;)

        X = []
        # Loop trough all label directories.
        for index, filename in dataframe[x_col].iteritems():
            x_i = self._transform_x(Path(filename))
            # Normalise estimate by exome size.
            x_i /= normalisation[index]

            # Concatenate other columns.
            if keep_columns:
                if x_i.index.nlevels &gt; 1:
                    raise ValueError(
                        &#34;Unable to coalesce single-index data frame with multi-index result from `transform`.&#34;
                    )
                if to_keep.size &gt; 0:
                    x_passthrough = dataframe.loc[[index], to_keep]
                    x_i.index = [index]
                    x_i = concat([x_i, x_passthrough], axis=&#34;columns&#34;)

            X.append(x_i)

        X_data_frame = concat(X, axis=&#34;rows&#34;)
        # Use index names from original data frame (instead of that given by
        # `transform`).
        if X_data_frame.index.nlevels &gt; 1:
            X_data_frame.index = X_data_frame.index.set_levels(dataframe.index, level=0)
        else:
            X_data_frame.index = dataframe.index

        if class_mode is None:
            return X_data_frame

        y = dataframe[y_col].copy()

        # Don&#39;t round when None.
        if decimals is None:
            return X_data_frame, y
        return X_data_frame.round(decimals), y

    def _transform_x(self, input_file: Path) -&gt; Series:
        &#34;&#34;&#34;Transform single record (e.g., a VCF or copy number file).&#34;&#34;&#34;
        return self.transform(input_file)</code></pre>
</details>
</section>
<section>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-classes">Classes</h2>
<dl>
<dt id="bearclaw.preprocessing.VariantDataGenerator"><code class="flex name class">
<span>class <span class="ident">VariantDataGenerator</span></span>
<span>(</span><span>transform: Callable)</span>
</code></dt>
<dd>
<div class="desc"><p>Transform (unstructured) mutation files to tabular format.</p>
<p>Generate tabular data from files containing variant or copy number calls.</p>
<h2 id="args">Args</h2>
<dl>
<dt><strong><code>transform</code></strong></dt>
<dd>Method that generates a row from a VCF or CNV file.</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">class VariantDataGenerator:
    &#34;&#34;&#34;Transform (unstructured) mutation files to tabular format.&#34;&#34;&#34;

    def __init__(
        self,
        transform: Callable,
    ):
        &#34;&#34;&#34;
        Generate tabular data from files containing variant or copy number calls.

        Args:
            transform: Method that generates a row from a VCF or CNV file.
        &#34;&#34;&#34;
        self.transform = transform

    def flow_from_dataframe(
        self,
        dataframe: DataFrame,
        x_col=&#34;filename&#34;,
        y_col=&#34;class&#34;,
        class_mode: Optional[Literal[&#34;raw&#34;]] = &#34;raw&#34;,
        keep_columns: bool = True,
        coverage_size: Optional[Union[float, str]] = None,
        decimals: Optional[int] = 2,
    ):
        &#34;&#34;&#34;Load dataset by reading VCF files and target label from dataframe.

        Args:
            dataframe: Pandas dataframe with columns pointing to VCF or CNV files.
            class_mode: When None, don&#39;t extract target label (inference mode).
            x_col: Column pointing to location of VCF or CNV file.
            y_col: Target label column.
            class_mode: Return features and labels during training mode (&#34;raw&#34;),
                or return only features during serving (None).
            keep_columns: Use only the features extracted through `x_col` (False) or
                also concatenate other columns in the dataframe after extraction (True).
            coverage_size: Normalise estimates by value (float), column (str), or not at
                all (None). Usually, this value is the size of the genomic region
                [typically in megabases (mb)] covered at sufficient depth to call
                variants.
            decimals: If not None, round to this many decimals.

        Returns: When class_mode is `None` return features `X`, otherwise return a
            pair (X, y) with labels `y`.

        &#34;&#34;&#34;
        if keep_columns:
            to_keep = dataframe.columns.difference([x_col, y_col])

        if coverage_size is None:
            normalisation = Series(1, index=dataframe.index)
        elif isinstance(coverage_size, (float, int)):
            normalisation = Series(coverage_size, index=dataframe.index)
        elif coverage_size in dataframe.columns:
            normalisation = dataframe[coverage_size]
        else:
            raise KeyError(f&#34;Unknown column {coverage_size}.&#34;)

        X = []
        # Loop trough all label directories.
        for index, filename in dataframe[x_col].iteritems():
            x_i = self._transform_x(Path(filename))
            # Normalise estimate by exome size.
            x_i /= normalisation[index]

            # Concatenate other columns.
            if keep_columns:
                if x_i.index.nlevels &gt; 1:
                    raise ValueError(
                        &#34;Unable to coalesce single-index data frame with multi-index result from `transform`.&#34;
                    )
                if to_keep.size &gt; 0:
                    x_passthrough = dataframe.loc[[index], to_keep]
                    x_i.index = [index]
                    x_i = concat([x_i, x_passthrough], axis=&#34;columns&#34;)

            X.append(x_i)

        X_data_frame = concat(X, axis=&#34;rows&#34;)
        # Use index names from original data frame (instead of that given by
        # `transform`).
        if X_data_frame.index.nlevels &gt; 1:
            X_data_frame.index = X_data_frame.index.set_levels(dataframe.index, level=0)
        else:
            X_data_frame.index = dataframe.index

        if class_mode is None:
            return X_data_frame

        y = dataframe[y_col].copy()

        # Don&#39;t round when None.
        if decimals is None:
            return X_data_frame, y
        return X_data_frame.round(decimals), y

    def _transform_x(self, input_file: Path) -&gt; Series:
        &#34;&#34;&#34;Transform single record (e.g., a VCF or copy number file).&#34;&#34;&#34;
        return self.transform(input_file)</code></pre>
</details>
<h3>Methods</h3>
<dl>
<dt id="bearclaw.preprocessing.VariantDataGenerator.flow_from_dataframe"><code class="name flex">
<span>def <span class="ident">flow_from_dataframe</span></span>(<span>self, dataframe: pandas.core.frame.DataFrame, x_col='filename', y_col='class', class_mode: Optional[Literal['raw']] = 'raw', keep_columns: bool = True, coverage_size: Union[float, str, ForwardRef(None)] = None, decimals: Optional[int] = 2)</span>
</code></dt>
<dd>
<div class="desc"><p>Load dataset by reading VCF files and target label from dataframe.</p>
<h2 id="args">Args</h2>
<dl>
<dt><strong><code>dataframe</code></strong></dt>
<dd>Pandas dataframe with columns pointing to VCF or CNV files.</dd>
<dt><strong><code>class_mode</code></strong></dt>
<dd>When None, don't extract target label (inference mode).</dd>
<dt><strong><code>x_col</code></strong></dt>
<dd>Column pointing to location of VCF or CNV file.</dd>
<dt><strong><code>y_col</code></strong></dt>
<dd>Target label column.</dd>
<dt><strong><code>class_mode</code></strong></dt>
<dd>Return features and labels during training mode ("raw"),
or return only features during serving (None).</dd>
<dt><strong><code>keep_columns</code></strong></dt>
<dd>Use only the features extracted through <code>x_col</code> (False) or
also concatenate other columns in the dataframe after extraction (True).</dd>
<dt><strong><code>coverage_size</code></strong></dt>
<dd>Normalise estimates by value (float), column (str), or not at
all (None). Usually, this value is the size of the genomic region
[typically in megabases (mb)] covered at sufficient depth to call
variants.</dd>
<dt><strong><code>decimals</code></strong></dt>
<dd>If not None, round to this many decimals.</dd>
</dl>
<p>Returns: When class_mode is <code>None</code> return features <code>X</code>, otherwise return a
pair (X, y) with labels <code>y</code>.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def flow_from_dataframe(
    self,
    dataframe: DataFrame,
    x_col=&#34;filename&#34;,
    y_col=&#34;class&#34;,
    class_mode: Optional[Literal[&#34;raw&#34;]] = &#34;raw&#34;,
    keep_columns: bool = True,
    coverage_size: Optional[Union[float, str]] = None,
    decimals: Optional[int] = 2,
):
    &#34;&#34;&#34;Load dataset by reading VCF files and target label from dataframe.

    Args:
        dataframe: Pandas dataframe with columns pointing to VCF or CNV files.
        class_mode: When None, don&#39;t extract target label (inference mode).
        x_col: Column pointing to location of VCF or CNV file.
        y_col: Target label column.
        class_mode: Return features and labels during training mode (&#34;raw&#34;),
            or return only features during serving (None).
        keep_columns: Use only the features extracted through `x_col` (False) or
            also concatenate other columns in the dataframe after extraction (True).
        coverage_size: Normalise estimates by value (float), column (str), or not at
            all (None). Usually, this value is the size of the genomic region
            [typically in megabases (mb)] covered at sufficient depth to call
            variants.
        decimals: If not None, round to this many decimals.

    Returns: When class_mode is `None` return features `X`, otherwise return a
        pair (X, y) with labels `y`.

    &#34;&#34;&#34;
    if keep_columns:
        to_keep = dataframe.columns.difference([x_col, y_col])

    if coverage_size is None:
        normalisation = Series(1, index=dataframe.index)
    elif isinstance(coverage_size, (float, int)):
        normalisation = Series(coverage_size, index=dataframe.index)
    elif coverage_size in dataframe.columns:
        normalisation = dataframe[coverage_size]
    else:
        raise KeyError(f&#34;Unknown column {coverage_size}.&#34;)

    X = []
    # Loop trough all label directories.
    for index, filename in dataframe[x_col].iteritems():
        x_i = self._transform_x(Path(filename))
        # Normalise estimate by exome size.
        x_i /= normalisation[index]

        # Concatenate other columns.
        if keep_columns:
            if x_i.index.nlevels &gt; 1:
                raise ValueError(
                    &#34;Unable to coalesce single-index data frame with multi-index result from `transform`.&#34;
                )
            if to_keep.size &gt; 0:
                x_passthrough = dataframe.loc[[index], to_keep]
                x_i.index = [index]
                x_i = concat([x_i, x_passthrough], axis=&#34;columns&#34;)

        X.append(x_i)

    X_data_frame = concat(X, axis=&#34;rows&#34;)
    # Use index names from original data frame (instead of that given by
    # `transform`).
    if X_data_frame.index.nlevels &gt; 1:
        X_data_frame.index = X_data_frame.index.set_levels(dataframe.index, level=0)
    else:
        X_data_frame.index = dataframe.index

    if class_mode is None:
        return X_data_frame

    y = dataframe[y_col].copy()

    # Don&#39;t round when None.
    if decimals is None:
        return X_data_frame, y
    return X_data_frame.round(decimals), y</code></pre>
</details>
</dd>
</dl>
</dd>
</dl>
</section>
</article>
<nav id="sidebar">
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="bearclaw" href="index.html">bearclaw</a></code></li>
</ul>
</li>
<li><h3><a href="#header-classes">Classes</a></h3>
<ul>
<li>
<h4><code><a title="bearclaw.preprocessing.VariantDataGenerator" href="#bearclaw.preprocessing.VariantDataGenerator">VariantDataGenerator</a></code></h4>
<ul class="">
<li><code><a title="bearclaw.preprocessing.VariantDataGenerator.flow_from_dataframe" href="#bearclaw.preprocessing.VariantDataGenerator.flow_from_dataframe">flow_from_dataframe</a></code></li>
</ul>
</li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc" title="pdoc: Python API documentation generator"><cite>pdoc</cite> 0.10.0</a>.</p>
</footer>
</body>
</html>