<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.10.0" />
<title>bearclaw.transforms API documentation</title>
<meta name="description" content="" />
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/sanitize.min.css" integrity="sha256-PK9q560IAAa6WVRRh76LtCaI8pjTJ2z11v0miyNNjrs=" crossorigin>
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/typography.min.css" integrity="sha256-7l/o7C8jubJiy74VsKTidCy1yBkRtiUGbVkYBylBqUg=" crossorigin>
<link rel="stylesheet preload" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/styles/github.min.css" crossorigin>
<style>:root{--highlight-color:#fe9}.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}#sidebar > *:last-child{margin-bottom:2cm}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}h1:target,h2:target,h3:target,h4:target,h5:target,h6:target{background:var(--highlight-color);padding:.2em 0}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{margin-top:.6em;font-weight:bold}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}dt:target .name{background:var(--highlight-color)}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}td{padding:0 .5em}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%;height:100vh;overflow:auto;position:sticky;top:0}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
<script async src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS_CHTML" integrity="sha256-kZafAc6mZvK3W3v1pHOcUix30OHQN6pU/NO2oFkqZVw=" crossorigin></script>
<script defer src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/highlight.min.js" integrity="sha256-Uv3H6lx7dJmRfRvH8TH6kJD1TSK1aFcwgx+mdg3epi8=" crossorigin></script>
<script>window.addEventListener('DOMContentLoaded', () => hljs.initHighlighting())</script>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>bearclaw.transforms</code></h1>
</header>
<section id="section-intro">
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">from pathlib import Path

from pandas import DataFrame, concat

from bearclaw.feature_extraction import CosmicNMF, _POSSIBLE_SBS_SEQUENCING_ARTEFACTS
from bearclaw.io import extract_mutation_spectra, from_tso500_trace, vcf


def spectrum(vcf_file: Path, exome: bool = False) -&gt; DataFrame:
    &#34;&#34;&#34;Extract single + doublet base substitutions and indel spectra from VCF file.&#34;&#34;&#34;
    mutation_matrices = extract_mutation_spectra(vcf_file, exome=exome)
    spectra = list(mutation_matrices.values())
    return concat(spectra, axis=&#34;columns&#34;)


@from_tso500_trace
def spectrum_from_trace(trace_file: Path) -&gt; DataFrame:
    &#34;&#34;&#34;Extract single + doublet base substitutions &amp; indel spectra from TSO500 trace.

    N.B. This spectrum incorporates all mutations that are eligible for a TSO500 TMB
    calculation (i.e., both synonymous and non-synonymous).

    Args:
        trace_file: Path to the *_TMB_Trace.tsv file generated by TSO500 app.
    &#34;&#34;&#34;
    return spectrum(trace_file, exome=False)


@vcf.non_synonymous
def non_synonymous_spectrum(vcf_file: Path, exome: bool = False) -&gt; DataFrame:
    return spectrum(vcf_file, exome)


def mutational_signature(
    vcf_file: Path, filter_seq_artefact_signatures: bool = True
) -&gt; DataFrame:
    &#34;&#34;&#34;
    Extract mutation spectra and deconvolute into COSMIC signatures.

    !! N.B.: Assumes that variants are from the GRCh37 reference genome. !!

    Args:
        vcf_file: File to analyse.
        filter_seq_artefact_signatures: When True, remove mutational
            signatures that are potentially related to sequencing artefacts.
    &#34;&#34;&#34;
    mutation_matrices = extract_mutation_spectra(vcf_file, exome=False)
    signatures = []
    for spectrum in [
        &#34;single_base_substitutions&#34;,
        &#34;doublet_base_substitutions&#34;,
        &#34;indel&#34;,
    ]:
        decomposer = CosmicNMF(cosmic_signature=spectrum)  # type: ignore
        signatures.append(decomposer.transform(mutation_matrices[spectrum]))

    result = concat(signatures, axis=&#34;columns&#34;)
    if filter_seq_artefact_signatures:
        return result.drop(columns=_POSSIBLE_SBS_SEQUENCING_ARTEFACTS)
    return result


@from_tso500_trace
def mutational_signature_from_trace(
    trace_file: Path, filter_seq_artefact_signatures: bool = True
) -&gt; DataFrame:
    &#34;&#34;&#34;Extract TSO500 mutation spectra and deconvolute into COSMIC signatures.

    N.B. This signature deconvolution uses all mutations that are eligible for a TMB
    calculation (i.e., both synonymous and non-synonymous). This is different from the
    true TMB definition (which includes only non-synonymous variants).

    Args:
        trace_file: Path to the *_TMB_Trace.tsv file generated by TSO500 app.
        filter_seq_artefact_signatures: When True, remove mutational
            signatures that are potentially related to sequencing artefacts.
    &#34;&#34;&#34;
    return mutational_signature(trace_file, filter_seq_artefact_signatures)</code></pre>
</details>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-functions">Functions</h2>
<dl>
<dt id="bearclaw.transforms.mutational_signature"><code class="name flex">
<span>def <span class="ident">mutational_signature</span></span>(<span>vcf_file: pathlib.Path, filter_seq_artefact_signatures: bool = True) ‑> pandas.core.frame.DataFrame</span>
</code></dt>
<dd>
<div class="desc"><p>Extract mutation spectra and deconvolute into COSMIC signatures.</p>
<p>!! N.B.: Assumes that variants are from the GRCh37 reference genome. !!</p>
<h2 id="args">Args</h2>
<dl>
<dt><strong><code>vcf_file</code></strong></dt>
<dd>File to analyse.</dd>
<dt><strong><code>filter_seq_artefact_signatures</code></strong></dt>
<dd>When True, remove mutational
signatures that are potentially related to sequencing artefacts.</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def mutational_signature(
    vcf_file: Path, filter_seq_artefact_signatures: bool = True
) -&gt; DataFrame:
    &#34;&#34;&#34;
    Extract mutation spectra and deconvolute into COSMIC signatures.

    !! N.B.: Assumes that variants are from the GRCh37 reference genome. !!

    Args:
        vcf_file: File to analyse.
        filter_seq_artefact_signatures: When True, remove mutational
            signatures that are potentially related to sequencing artefacts.
    &#34;&#34;&#34;
    mutation_matrices = extract_mutation_spectra(vcf_file, exome=False)
    signatures = []
    for spectrum in [
        &#34;single_base_substitutions&#34;,
        &#34;doublet_base_substitutions&#34;,
        &#34;indel&#34;,
    ]:
        decomposer = CosmicNMF(cosmic_signature=spectrum)  # type: ignore
        signatures.append(decomposer.transform(mutation_matrices[spectrum]))

    result = concat(signatures, axis=&#34;columns&#34;)
    if filter_seq_artefact_signatures:
        return result.drop(columns=_POSSIBLE_SBS_SEQUENCING_ARTEFACTS)
    return result</code></pre>
</details>
</dd>
<dt id="bearclaw.transforms.mutational_signature_from_trace"><code class="name flex">
<span>def <span class="ident">mutational_signature_from_trace</span></span>(<span>trace_file: pathlib.Path, filter_seq_artefact_signatures: bool = True) ‑> pandas.core.frame.DataFrame</span>
</code></dt>
<dd>
<div class="desc"><p>Extract TSO500 mutation spectra and deconvolute into COSMIC signatures.</p>
<p>N.B. This signature deconvolution uses all mutations that are eligible for a TMB
calculation (i.e., both synonymous and non-synonymous). This is different from the
true TMB definition (which includes only non-synonymous variants).</p>
<h2 id="args">Args</h2>
<dl>
<dt><strong><code>trace_file</code></strong></dt>
<dd>Path to the *_TMB_Trace.tsv file generated by TSO500 app.</dd>
<dt><strong><code>filter_seq_artefact_signatures</code></strong></dt>
<dd>When True, remove mutational
signatures that are potentially related to sequencing artefacts.</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">@from_tso500_trace
def mutational_signature_from_trace(
    trace_file: Path, filter_seq_artefact_signatures: bool = True
) -&gt; DataFrame:
    &#34;&#34;&#34;Extract TSO500 mutation spectra and deconvolute into COSMIC signatures.

    N.B. This signature deconvolution uses all mutations that are eligible for a TMB
    calculation (i.e., both synonymous and non-synonymous). This is different from the
    true TMB definition (which includes only non-synonymous variants).

    Args:
        trace_file: Path to the *_TMB_Trace.tsv file generated by TSO500 app.
        filter_seq_artefact_signatures: When True, remove mutational
            signatures that are potentially related to sequencing artefacts.
    &#34;&#34;&#34;
    return mutational_signature(trace_file, filter_seq_artefact_signatures)</code></pre>
</details>
</dd>
<dt id="bearclaw.transforms.non_synonymous_spectrum"><code class="name flex">
<span>def <span class="ident">non_synonymous_spectrum</span></span>(<span>vcf_file: pathlib.Path, exome: bool = False) ‑> pandas.core.frame.DataFrame</span>
</code></dt>
<dd>
<div class="desc"></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">@vcf.non_synonymous
def non_synonymous_spectrum(vcf_file: Path, exome: bool = False) -&gt; DataFrame:
    return spectrum(vcf_file, exome)</code></pre>
</details>
</dd>
<dt id="bearclaw.transforms.spectrum"><code class="name flex">
<span>def <span class="ident">spectrum</span></span>(<span>vcf_file: pathlib.Path, exome: bool = False) ‑> pandas.core.frame.DataFrame</span>
</code></dt>
<dd>
<div class="desc"><p>Extract single + doublet base substitutions and indel spectra from VCF file.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def spectrum(vcf_file: Path, exome: bool = False) -&gt; DataFrame:
    &#34;&#34;&#34;Extract single + doublet base substitutions and indel spectra from VCF file.&#34;&#34;&#34;
    mutation_matrices = extract_mutation_spectra(vcf_file, exome=exome)
    spectra = list(mutation_matrices.values())
    return concat(spectra, axis=&#34;columns&#34;)</code></pre>
</details>
</dd>
<dt id="bearclaw.transforms.spectrum_from_trace"><code class="name flex">
<span>def <span class="ident">spectrum_from_trace</span></span>(<span>trace_file: pathlib.Path) ‑> pandas.core.frame.DataFrame</span>
</code></dt>
<dd>
<div class="desc"><p>Extract single + doublet base substitutions &amp; indel spectra from TSO500 trace.</p>
<p>N.B. This spectrum incorporates all mutations that are eligible for a TSO500 TMB
calculation (i.e., both synonymous and non-synonymous).</p>
<h2 id="args">Args</h2>
<dl>
<dt><strong><code>trace_file</code></strong></dt>
<dd>Path to the *_TMB_Trace.tsv file generated by TSO500 app.</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">@from_tso500_trace
def spectrum_from_trace(trace_file: Path) -&gt; DataFrame:
    &#34;&#34;&#34;Extract single + doublet base substitutions &amp; indel spectra from TSO500 trace.

    N.B. This spectrum incorporates all mutations that are eligible for a TSO500 TMB
    calculation (i.e., both synonymous and non-synonymous).

    Args:
        trace_file: Path to the *_TMB_Trace.tsv file generated by TSO500 app.
    &#34;&#34;&#34;
    return spectrum(trace_file, exome=False)</code></pre>
</details>
</dd>
</dl>
</section>
<section>
</section>
</article>
<nav id="sidebar">
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="bearclaw" href="index.html">bearclaw</a></code></li>
</ul>
</li>
<li><h3><a href="#header-functions">Functions</a></h3>
<ul class="">
<li><code><a title="bearclaw.transforms.mutational_signature" href="#bearclaw.transforms.mutational_signature">mutational_signature</a></code></li>
<li><code><a title="bearclaw.transforms.mutational_signature_from_trace" href="#bearclaw.transforms.mutational_signature_from_trace">mutational_signature_from_trace</a></code></li>
<li><code><a title="bearclaw.transforms.non_synonymous_spectrum" href="#bearclaw.transforms.non_synonymous_spectrum">non_synonymous_spectrum</a></code></li>
<li><code><a title="bearclaw.transforms.spectrum" href="#bearclaw.transforms.spectrum">spectrum</a></code></li>
<li><code><a title="bearclaw.transforms.spectrum_from_trace" href="#bearclaw.transforms.spectrum_from_trace">spectrum_from_trace</a></code></li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc" title="pdoc: Python API documentation generator"><cite>pdoc</cite> 0.10.0</a>.</p>
</footer>
</body>
</html>